File: Python/ProjectEuler_14.py
# Longest Collatz Sequence -- Solved
# Solution = 837799

from Timer import Timer
Stopwatch = Timer()

Solution = (0,0)
for i in range(1000000):
    n = i
    CollatzSequence = [n]

    while n > 1:
        if n % 2 == 0:
            n //= 2
        else:
            n = 3*n+1
        CollatzSequence.append(n)

    if len(CollatzSequence) > Solution[1]:
        Solution = (i, len(CollatzSequence))

print('The starting number under one million that produces the largest Collatz sequence is : ', Solution[0])
Stopwatch.stop()

File: MATLAB/ProjectEuler_14.m
%% Longest Collatz Sequence -- Solved
clear, close, clc, disp(mfilename), tic;

%% Solution:
Limit = 1000000;
CollatzSequences = [1:Limit]';

for i = 1:Limit
    n = i;
    for j = 2:Limit^2
        if n == 1
            break;
        end
        
        p = Collatz(n);
        n = p;
        CollatzSequences(i,j) = p;
    end
end

a = CollatzSequences';
b = a > 0;
c = sum(b);

Solution = find(c == max(c));

%% Display Solution
fprintf(['The starting number under one million that produces the largest ' ...
    'Collatz sequence is : %d\n'], Solution); toc;

%% Collatz()
function n2 = Collatz(n1)
    if rem(n1,2) == 0
        n2 = n1/2;
    else
        n2 = 3*n1+1;
    end
end